Library/Homebrew/utils/ruby.sh

```diff
@@ -6,17 +13,23 @@ setup-ruby-path() {
   # When bumping check if HOMEBREW_MACOS_SYSTEM_RUBY_NEW_ENOUGH (in brew.sh)
   # also needs to be changed.
   local required_ruby_version="2.6"
+  local old_ruby_path ruby_clean_path ruby_user_path
+  local advice="
+If there's no Homebrew Portable Ruby available for your processor:
+- install Ruby $required_ruby_version with your system package manager (or rbenv/ruby-build)
+- make it first in your PATH
+- try again
+"
 
   vendor_dir="$HOMEBREW_LIBRARY/Homebrew/vendor"
   vendor_ruby_current_version="$vendor_dir/portable-ruby/current"
   vendor_ruby_path="$vendor_ruby_current_version/bin/ruby"
 
-  if [[ -z "$HOMEBREW_DEVELOPER" ]]
-  then
-    unset HOMEBREW_RUBY_PATH
-  fi
+  [[ -n $HOMEBREW_RUBY_PATH ]] && old_ruby_path=$HOMEBREW_RUBY_PATH
```

Comment:
I'm probably missing something but: why keep this around at all if we're checking `PATH` and `HOMEBREW_PATH` anyway?

Reply:
This part handles the cases when user sets `HOMEBREW_RUBY_PATH` directly. Ability to set `HOMEBREW_RUBY_PATH` directly instead of modifying `PATH` is congruent with how Homebrew allows to set `HOMEBREW_CURL_PATH` and `HOMEBREW_GIT_PATH`.

Reply:
> Homebrew allows to set `HOMEBREW_CURL_PATH` and `HOMEBREW_GIT_PATH`.

We don't allow that (intentionally at least) and don't document them as supported. `HOMEBREW_FORCE_BREWED_CURL` and `HOMEBREW_FORCE_BREWED_GIT` are the entry points for that. What's the use-case for not wanting to use the `ruby` in your `HOMEBREW_PATH` or `PATH` or the vendored Ruby.

Reply:
I don't see this applicable in Mac but in linux
i have a ruby 2.6 installed just to make homebrew work, the ruby version provided by the package manager is 2.7. I don't want the homebrew ruby in PATH. 

Of course this is not a strict requirement, i can chuck away the homebrew ruby somewhere behind in PATH. 
but this implementation (`type -P ruby`) finds the first executable ruby, and it will fail and ask me to install ruby 2.6(or try to download) even though a ruby 2.6 is present.

Reply:
I see. I guess ideally we'd actually search through the `HOMEBREW_PATH` and `PATH` (`which -a`?) to find a `ruby` that passes our test and only give up if we don't find any?

Reply:
> `HOMEBREW_FORCE_BREWED_CURL` and `HOMEBREW_FORCE_BREWED_GIT` are the entry points for that.

Currently, `HOMEBREW_CURL_PATH` and `HOMEBREW_GIT_PATH` are critical for getting Homebrew up and running on systems where cURL and Git don't satisfy Homebrew requirements. I think it makes sense to make `HOMEBREW_RUBY_PATH` variable behave the same way. We can certainly parse the entire `PATH` and find the right one if it exists.

`HOMEBREW_FORCE_BREWED_*` are in a different category and don't help in situations where Homebrew won't even run. 

> ideally we'd actually search through the `HOMEBREW_PATH` and `PATH` (`which -a`?) to find a `ruby` that passes our test and only give up if we don't find any?

I will implement that.

Reply:
> Currently, `HOMEBREW_CURL_PATH` and `HOMEBREW_GIT_PATH` are critical for getting Homebrew up and running on systems where cURL and Git don't satisfy Homebrew requirements.

I don't agree. `brew update` has code specifically to install `curl` and `git` if needed before doing anything else. If that doesn't work: let's fix it rather than relying on hacking undocumented functionality to make it work.

> `HOMEBREW_FORCE_BREWED_*` are in a different category and don't help in situations where Homebrew won't even run.

They are set automatically for the situations where Homebrew cannot run. If they aren't: that's a bug we should fix.

Reply:
> I don't agree.

You can disagree but I'm speaking from experience of installing + making Homebrew work on a system with old system's ~everything~ curl and git and their modern versions in PATH. Anyway, let's finalize this PR first and then fix other things.

Reply:
Ruby is now searched in the default and user PATHs only. HOMEBREW_RUBY_PATH has no effect. I've tested this on the system with old system Ruby and it works! 🎉

Reply:
I have to say that searching through the clean and user PATHs is likely to impact performance. That `old_ruby_path` was playing a role of cache, I guess.

Reply:
> I don't agree. brew update has code specifically to install curl and git if needed before doing anything else. If that doesn't work: let's fix it rather than relying on hacking undocumented functionality to make it work.

If the system `curl` is too old, and the user has a newer version of `curl` say in their Home directory, `HOMEBREW_CURL_PATH` is currently how you point Homebrew to that newer version of `curl`. Homebrew can't download the bottles or source code of `curl` without a recent version of `curl`, so `HOMEBREW_FORCE_BREWED_CURL` is not helpful in this case.

Reply:
`HOMEBREW_CURL_PATH` are is the intentionally developer-exposed environment variable, and `HOMEBREW_CURL` is the not developer-exposed internal enivronment variable. `HOMEBREW_CURL_PATH ` is active only when `HOMEBREW_DEVELOPER` is active. See https://github.com/Homebrew/brew/blob/b2fc8ad7b3fa874ac76f4ee2fba48f7535487caa/Library/Homebrew/brew.sh#L96-L106

Reply:
> I have to say that searching through the clean and user PATHs is likely to impact performance. That `old_ruby_path` was playing a role of cache, I guess.

I ran few tests,

#### ruby 2.6 is first in `which -a ruby`
![first-in-path](https://user-images.githubusercontent.com/8173131/82395997-72b69280-9a6a-11ea-9684-a15c5f9691b2.png)

#### ruby 2.6 is last in  `which -a ruby`
![last-in-path](https://user-images.githubusercontent.com/8173131/82396040-92e65180-9a6a-11ea-8950-3983e514f268.png)

#### ruby 2.6 set using HOMEBREW_RUBY_PATH(tested on previous commit)
![hb-rb-path](https://user-images.githubusercontent.com/8173131/82396173-e062be80-9a6a-11ea-9418-eda90368130c.png)


once the 2.6 executable is found, may be we cache it somewhere. that should give a little bump in the next runs. 
Searching through PATH for every `brew` call don't seem such a nice idea imo.

Reply:
> so `HOMEBREW_FORCE_BREWED_CURL` is not helpful in this case.

I thought `curl` had an HTTP mirror but I forgot it no longer does as I did not know this was required for Linux: https://github.com/Homebrew/homebrew-core/commit/2f5aa1443267d0027c0370a5620d4423c886d6fa. There was also an `audit` to ensure that `curl` always had an HTTP mirror.

I'd suggest that if we want to be more accessible to older Linux versions we go down that route rather than requiring the user to manually find and download a suitable `curl`, set `HOMEBREW_DEVELOPER` and use an undocumented and unsupported environment variable.

Alternatively or additionally, we could take the same approach as here with `curl` and `git` where we look through all of those in the `PATH`.
> once the 2.6 executable is found, may be we cache it somewhere.

This seems like a nice idea. Would be cool to have in this PR but I don't consider it blocking; this is already a fairly niche case (which we could warn users on?).

Reply:
> > once the 2.6 executable is found, may be we cache it somewhere.
> 
> This seems like a nice idea. Would be cool to have in this PR but I don't consider it blocking; this is already a fairly niche case (which we could warn users on?).

I can add it in this PR.